Configurable string encodings (#133)

Joshhua5 opened the pull request:
Added the ability to optionally override the encoding on string operations from Memory.

Joshhua5 committed:
Configurable string encodings (#133)

peterhuene commented on src/Memory.cs:
```diff
@@ -62,6 +62,11 @@ public Memory(IStore store, uint minimum = 0, uint maximum = uint.MaxValue)
+        /// <summary>
+        /// The memory encoding for string operations. 
+        /// </summary>
+        public Encoding MemoryEncoding { get; set; } = Encoding.UTF8;
```
Would perhaps a `Encoding` parameter to `ReadString` and `WriteString` defaulted to `Encoding.UTF8` would be a little more ergonomic for anyone wishing to read different string encodings from the same linear memory (although not as likely as scenario)?

peterhuene replied:
Related, we should update the doc comments for `ReadString` and `WriteString` so as to not be UTF-8 specific.

peterhuene commented on src/Memory.cs:
```diff
@@ -131,7 +136,7 @@ public string ReadNullTerminatedString(IStore store, int address)
             }
 
-            return Encoding.UTF8.GetString(slice.Slice(0, terminator));
+            return MemoryEncoding.GetString(slice.Slice(0, terminator));
```
I think we should keep `ReadNullTerminatedString` specific to UTF-8 given that we can't can't simply `IndexOf` for the null byte in other encodings.

peterhuene reviewed (commented):
Hi @Joshhua5. Thanks for putting this PR up!

Just a few review comments to go over.

peterhuene commented:
Hi @Joshhua5! I just wanted to touch base and see if you saw the feedback to your PR.

Commit:
Make ReadString and WriteString take an Encoding parameter.

This commit updates Josh's work to make `Memory.ReadString` and
`Memory.WriteString` accept a specified encoding rather than always defaulting
to UTF-8.

A few additional files have been formatted as well.

peterhuene commented:
Hi @Joshhua5. As I haven't heard from you in a while, I went ahead and amended your PR to include the requested changes.

As you did not create a local branch for your changes, I unfortunately had to push to your `main` branch to update the PR.

When this PR merges, please run `git fetch upstream && git reset --hard upstream/main` to hard reset your main branch to upstream. This assumes you've set up a remote named `upstream` for `git@github.com:bytecodealliance/wasmtime-dotnet.git`, though.

peterhuene commented:
@Joshhua5 thanks again for working on these changes!

Joshhua5 commented:
@peterhuene Thank you for following through with this PR, it would be good for me to have notifications turn on.
